File: leetcode/fruit_basket.py
# Input: fruits = [1,2,1]
# Output: 3
# Explanation: We can pick from all 3 trees.

# Input: fruits = [1,2,3,2,2]
# Output: 4
# Explanation: We can pick from trees [2,3,2,2].


class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        count, i = {}, 0
        for j, v in enumerate(fruits):
            count[v] = count.get(v, 0) + 1
            if len(count) > 2:
                count[fruits[i]] -= 1
                if count[fruits[i]] == 0:
                    del count[fruits[i]]
                i += 1
        return j - i + 1

File: leetcode/fruit_basket.ts
// Input: fruits = [0,1,2,2]
// Output: 3

// basically a sliding window problem with a terrible description

function totalFruit(fruits: number[]): number {
  if (!fruits || fruits.length === 0) return 0;
  const isSeens = {};
  let isSeenCount = 0;
  let maxCount = 0;
  let left = 0,
    right = 0;
  while (right < fruits.length) {
    const char = fruits[right];
    if (isSeens[char] !== undefined) {
      isSeens[char]++;
    } else {
      isSeens[char] = 1;
      isSeenCount++;
      while (isSeenCount > 2) {
        const lChar = fruits[left];
        isSeens[lChar]--;
        left++;
        if (isSeens[lChar] === 0) {
          delete isSeens[lChar];
          isSeenCount--;
        }
      }
    }
    maxCount = Math.max(maxCount, right - left + 1);
    right++;
  }
  return maxCount;
}
